File: src/openagents/repository/task_repository.py
from datetime import datetime

from sqlalchemy import select, delete
from sqlalchemy.orm import selectinload

from openagents.repository.database import ConversationEntity, TaskEntity, async_session


# 查询全部任务，按 id 升序
async def list_tasks() -> list[TaskEntity]:
    async with async_session() as session:
        result = await session.execute(select(TaskEntity).order_by(TaskEntity.id))
        return list(result.scalars().all())


# 按 id 查询任务，预加载阶段对话（含每条对话的消息与执行 Agent），不存在返回 None
async def get_task(task_id: int) -> TaskEntity | None:
    async with async_session() as session:
        result = await session.execute(
            select(TaskEntity)
            .where(TaskEntity.id == task_id)
            .options(
                selectinload(TaskEntity.conversations).selectinload(ConversationEntity.messages),
                selectinload(TaskEntity.conversations).selectinload(ConversationEntity.agent),
            )
        )
        task = result.scalar_one_or_none()
        if task is None:
            return None
        task.conversations.sort(key=lambda x: x.id)
        for conversation in task.conversations:
            conversation.messages.sort(key=lambda message: message.id)
        return task


# 新增任务，时间字段统一赋当前时间，返回自增 id
async def add_task(title: str, content: str, agent_ids: list[int], work_dir: str) -> int:
    async with async_session() as session:
        now = datetime.now()
        task = TaskEntity(title=title, content=content, agent_ids=agent_ids, work_dir=work_dir, create_time=now, update_time=now)
        session.add(task)
        await session.commit()
        await session.refresh(task)
        return task.id


# 按 id 删除任务，不存在返回 False，关联对话由数据库外键 ON DELETE CASCADE 级联删除
async def delete_task(task_id: int) -> bool:
    async with async_session() as session:
        result = await session.execute(delete(TaskEntity).where(TaskEntity.id == task_id))
        await session.commit()
        return result.rowcount > 0

File: src/repository/task_repository.rs
// 任务 CRUD
use sqlx::SqlitePool;

use super::entity::TaskEntity;

// 查询全部任务, 按 id 升序
pub async fn list_tasks(pool: &SqlitePool) -> Result<Vec<TaskEntity>, sqlx::Error> {
    sqlx::query_as::<_, TaskEntity>(
        "SELECT id, title, content, agent_ids, work_dir, create_time, update_time FROM t_task ORDER BY id",
    )
        .fetch_all(pool)
        .await
}

// 按 id 查询任务基本字段
pub async fn get_task(pool: &SqlitePool, task_id: i64) -> Result<Option<TaskEntity>, sqlx::Error> {
    sqlx::query_as::<_, TaskEntity>(
        "SELECT id, title, content, agent_ids, work_dir, create_time, update_time FROM t_task WHERE id = ?",
    )
        .bind(task_id)
        .fetch_optional(pool)
        .await
}

// 新增任务, 返回自增 id
pub async fn add_task(
    pool: &SqlitePool,
    title: &str,
    content: &str,
    agent_ids: &[i64],
    work_dir: &str,
) -> Result<i64, sqlx::Error> {
    let now = chrono::Local::now().to_rfc3339();
    let result = sqlx::query(
        "INSERT INTO t_task (title, content, agent_ids, work_dir, create_time, update_time) VALUES (?, ?, ?, ?, ?, ?)",
    )
        .bind(title)
        .bind(content)
        .bind(sqlx::types::Json(agent_ids))
        .bind(work_dir)
        .bind(&now)
        .bind(&now)
        .execute(pool)
        .await?;
    Ok(result.last_insert_rowid())
}

// 按 id 更新任务, 不存在返回 false
pub async fn update_task(
    pool: &SqlitePool,
    task_id: i64,
    title: &str,
    content: &str,
    agent_ids: &[i64],
    work_dir: &str,
) -> Result<bool, sqlx::Error> {
    let now = chrono::Local::now().to_rfc3339();
    let result = sqlx::query(
        "UPDATE t_task SET title = ?, content = ?, agent_ids = ?, work_dir = ?, update_time = ? WHERE id = ?",
    )
        .bind(title)
        .bind(content)
        .bind(sqlx::types::Json(agent_ids))
        .bind(work_dir)
        .bind(&now)
        .bind(task_id)
        .execute(pool)
        .await?;
    Ok(result.rows_affected() > 0)
}

// 按 id 删除任务, 不存在返回 false
pub async fn delete_task(pool: &SqlitePool, task_id: i64) -> Result<bool, sqlx::Error> {
    let result = sqlx::query("DELETE FROM t_task WHERE id = ?")
        .bind(task_id)
        .execute(pool)
        .await?;
    Ok(result.rows_affected() > 0)
}
